DOCSP-37002 - GridFS (#18)
mongodb/docs-pymongo: source/fundamentals/gridfs.txt

.. _pymongo-gridfs:

GridFS
======

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: binary large object, blob, file, storage

> This guide shows how to use GridFS to store and retrieve large
> in MongoDB.

Reviewer: ```suggestion
This guide shows how to use GridFS to store and retrieve large files
in MongoDB.
```

Author: good catch.

Setup
-----

> Each GridFS instance operates on a specific MongoDB database.

Reviewer: S: monospace since this seems to be talking about the GridFS instance rather that just GridFS in general

```suggestion
Each ``GridFS`` instance operates on a specific MongoDB database.
```

To create a ``GridFS`` instance, call the ``GridFS()``
constructor and pass in a MongoDB database:

.. code-block:: python

   >>> from pymongo import MongoClient
   >>> import gridfs
   >>>
   >>> db = MongoClient().gridfs_example
   >>> fs = gridfs.GridFS(db)

Saving and Retrieving Data
--------------------------

GridFS offers a key-value-pair interface through the ``~gridfs.GridFS.put()`` and
``~gridfs.GridFS.get()`` methods.

To write data to GridFS, use the ``~gridfs.GridFS.put()`` method:

.. code-block:: python

   >>> a = fs.put("hello world")

The ``~gridfs.GridFS.put()`` method creates a new file in GridFS, and returns

> the value of the file document's ``_id`` key. You can use the file's returned ``_id``

Reviewer: Q: Is "file document" a GridFS term? Or should it be either "file" or "document" (referring to the "value of the file document's ... " part here)?

Author: Good question. This is the language of the original page, but our other docs just say "file", so that's what I'll use here.

value to retrieve the contents of the file:

.. code-block:: python

   >>> fs.get(a).read()
   'hello world'

The ``~gridfs.GridFS.get()`` method returns a file-like object. To read the

> file's contents, call the ``~gridfs.grid_file.GridOut.read()`` method. You can store
> any object with a ``read()`` method in GridFS.

Reviewer: S: I was initially confused because I interpreted this as saying to use a ``read()`` method to store the object

```suggestion
file's contents, call the ``~gridfs.grid_file.GridOut.read()`` method. You can store
any object that has a ``read()`` method in GridFS.
```
